fondant/dataset.py

```diff
@@ -76,60 +105,124 @@ def load_dataframe(self, spec: FondantComponentSpec) -> dd.DataFrame:
 
         return df
 
-    def _upload_index(self, df: dd.DataFrame):
-        # get remote path
-        remote_path = self.manifest.index.location
-
-        # upload to the cloud
-        dd.to_parquet(
-            df,
-            remote_path,
-            schema=self.index_schema,
-            overwrite=True,
+    @staticmethod
+    def _create_write_dataframe_task(
+        *, df: dd.DataFrame, remote_path: str, schema: t.Dict[str, str]
+    ) -> dd.core.Scalar:
+        """
+        Creates a delayed Dask task to upload the given DataFrame to the remote storage location
+         specified in the manifest.
+        Args:
+            df: The DataFrame to be uploaded.
+            remote_path: the location to upload the subset to
+            schema: the schema of the dataframe to write
+        Returns:
+             A delayed Dask task that uploads the DataFrame to the remote storage location when
+              executed.
+        """
+
+        # Define task to upload index to remote storage
+        upload_index_task = dd.to_parquet(
+            df, remote_path, schema=schema, overwrite=True, compute=False
         )
-
-    def _upload_subset(
-        self, name: str, fields: t.Mapping[str, Field], df: dd.DataFrame
-    ):
-        # create expected schema
-        expected_schema = {field.name: field.type.name for field in fields.values()}
-        expected_schema.update(self.index_schema)
-
-        # get remote path
-        remote_path = self.manifest.subsets[name].location
-        # upload to the cloud
-        dd.to_parquet(df, remote_path, schema=expected_schema, overwrite=True)
+        return upload_index_task
 
     def write_index(self, df: dd.DataFrame):
+        """
+        Write the index dataframe to a remote location.
+        Args:
+            df: The input Dask dataframe.
+        """
+        remote_path = self.manifest.index.location
         index_columns = list(self.manifest.index.fields.keys())
 
         # load index dataframe
         index_df = df[index_columns]
 
-        self._upload_index(index_df)
+        upload_index_task = self._create_write_dataframe_task(
+            df=index_df, remote_path=remote_path, schema=self.index_schema
+        )
+
+        # Write index
+        dd.compute(upload_index_task)
 
     def write_subsets(self, df: dd.DataFrame, spec: FondantComponentSpec):
-        for name, subset in spec.output_subsets.items():
-            fields = list(subset.fields.keys())
-            # verify fields are present in the output dataframe
-            subset_columns = [f"{name}_{field}" for field in fields]
+        """
+        Write subset tasks to a remote location
+
+        Args:
+            df (dask.dataframe.DataFrame): The input Dask dataframe.
+            spec (FondantComponentSpec): The specification of the output subsets.
+
+        Raises:
+            ValueError: If a field defined in an output subset is not present in the input
```

Comment:
"input dataframe" is a bit confusing here, as it could refer to the dataframe we provide as the input of the component as well.

Reply:
I will just call it `dataframe` maybe to avoid confusion

Reply:
or `output dataframe` as you specified below might be better